packages/core/src/components/overlay/overlay.tsx

```diff
@@ -348,8 +349,18 @@ export class Overlay extends React.PureComponent<IOverlayProps, IOverlayState> {
     private handleDocumentClick = (e: MouseEvent) => {
         const { isOpen, onClose } = this.props;
         const eventTarget = e.target as HTMLElement;
+
+        const { openStack } = Overlay;
+        const stackIndex = openStack.indexOf(this);
+
+        const isClickInDescendantOverlay = openStack
+            .slice(stackIndex)
+            .map(findDOMNode)
+            .some(elem => elem && elem.contains && elem.contains(eventTarget));
+
         const isClickInOverlay = this.containerElement != null && this.containerElement.contains(eventTarget);
-        if (isOpen && this.props.canOutsideClickClose && !isClickInOverlay) {
+
+        if (isOpen && this.props.canOutsideClickClose && !isClickInOverlay && !isClickInDescendantOverlay) {
```

Comment:
Since `isClickInDescendantOverlay` includes _this_ overlay instance (`openStack.slice(stackIndex)` will include this Overlay), I think we can remove `isClickInOverlay` altogether.